### PR TITLE
Initial basic implementation

### PR DESCRIPTION
**Pull request by asgrim**

### CODE REVIEW

File: src/Comparator.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use Roave\BetterReflection\Reflector\ClassReflector;
+use Roave\BetterReflection\Reflector\Exception\IdentifierNotFound;
+
+final class Comparator
+{
+    public function compare(ClassReflector $oldApi, ClassReflector $newApi): array
+    {
+        $changelog = [];
+
+        foreach ($oldApi->getAllClasses() as $oldClass) {
```
Comment (Ocramius) [inline src/Comparator.php]: To be isolated into a method first

File: src/Comparator.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            try {
+                $newClass = $newApi->reflect($oldClass->getName());
```
Comment (Ocramius) [inline src/Comparator.php]: This approach only looks for BC breaks, not API additions

File: src/Comparator.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            } catch (IdentifierNotFound $exception) {
+                $changelog[] = sprintf('[BC] Class %s has been deleted', $oldClass->getName());
+                continue;
+            }
+
+            foreach ($oldClass->getMethods() as $oldMethod) {
```
Comment (Ocramius) [inline src/Comparator.php]: To be isolated into a private method

File: src/Comparator.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                } catch (\OutOfBoundsException $exception) {
+                    $changelog[] = sprintf(
+                        '[BC] Method %s in class %s has been deleted',
+                        $oldMethod->getName(),
+                        $oldClass->getName()
+                    );
+                    continue;
+                }
+
+                foreach ($oldMethod->getParameters() as $oldParameter) {
```
Comment (Ocramius) [inline src/Comparator.php]: And again

**Comment by asgrim** [comment]
Jeez man, gimme a break, I literally just started this :P

File: src/Comparator.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    $newParameter = $newMethod->getParameter($oldParameter->getName());
```
Comment (Ocramius) [inline src/Comparator.php]: The parameter should not be fetched by name, but by position. A simple test of a BC compliant change is:

before:
```php
function foo(int $a, string $b) {}
```

after:
```php
function foo(int $b, string $a) {}
```

Also, a rename per-se does not cause a BC break:


before:
```php
function foo(int $a, string $b) {}
```

after:
```php
function foo(int $a, string $c) {}
```

**Comment by Ocramius** [comment]
> gimme a break
rageface

File: src/Comparator.php (hunk 2 above)
Comment (samdark) [reply src/Comparator.php]: As I understand, this one is to look for public API removal, see `catch`.

**Comment by asgrim** [comment]
@samdark thanks for your comments; it's still a little way off yet, hopefully I'll find time soon to work on this again :)

File: src/Comparator.php [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+use Roave\BetterReflection\Reflection\ReflectionClass;
+use Roave\BetterReflection\Reflection\ReflectionMethod;
+use Roave\BetterReflection\Reflection\ReflectionParameter;
+use Roave\BetterReflection\Reflector\ClassReflector;
+use Roave\BetterReflection\Reflector\Exception\IdentifierNotFound;
+
+final class Comparator
+{
+    public function compare(ClassReflector $oldApi, ClassReflector $newApi): array
```
Comment (Ocramius) [inline src/Comparator.php]: Do we only compare classes?

File: src/Comparator.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        $changelog = [];
+
+        foreach ($oldApi->getAllClasses() as $oldClass) {
+            $changelog = $this->examineClass($changelog, $oldClass, $newApi);
+        }
+
+        return $changelog;
+    }
+
+    private function examineClass(array $changelog, ReflectionClass $oldClass, ClassReflector $newApi): array
```
Comment (Ocramius) [inline src/Comparator.php]: Can we avoid returning `array`?

File: src/Comparator.php (hunk 6 above)
Comment (Ocramius) [reply src/Comparator.php]: Return type `array` is too limited

File: src/Comparator.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+        return $changelog;
+    }
+
+    private function examineMethod(
+        array $changelog,
+        ReflectionClass $oldClass,
+        ReflectionMethod $oldMethod,
+        ReflectionClass $newClass
+    ): array {
```
Comment (Ocramius) [inline src/Comparator.php]: Same here: `array` is too limited

File: src/Comparator.php (hunk 6 above)
Comment (asgrim) [reply src/Comparator.php]: For now yes, if you want to support functions, add an issue :P

**Comment by asgrim** [comment]
Created #13 to discuss the actual API to be used further :+1: for now merging so other functionality can start to be added

**Pull request merged**